From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Sat, 17 May 2025 18:20:04 +0200
Subject: [PATCH 088/225] dt-bindings: media: Add bindings for Unisoc camera
 subsystem
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Add device tree bindings for the camera subsystem found in the UMS9230
SoC.

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 .../bindings/media/sprd,camsys.yaml           | 366 ++++++++++++++++++
 include/dt-bindings/media/sprd-camsys.h       |  19 +
 2 files changed, 385 insertions(+)
 create mode 100644 Documentation/devicetree/bindings/media/sprd,camsys.yaml
 create mode 100644 include/dt-bindings/media/sprd-camsys.h

diff --git a/Documentation/devicetree/bindings/media/sprd,camsys.yaml b/Documentation/devicetree/bindings/media/sprd,camsys.yaml
new file mode 100644
index 000000000..40c4ec252
--- /dev/null
+++ b/Documentation/devicetree/bindings/media/sprd,camsys.yaml
@@ -0,0 +1,366 @@
+# SPDX-License-Identifier: (GPL-2.0 OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/media/sprd,camsys.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: Unisoc camera subsystem
+
+maintainers:
+  - Otto Pflüger <otto.pflueger@abscue.de>
+
+description:
+  The camera subsystem found in Unisoc SoCs consists of multiple CSI PHYs, CSI
+  controllers and DCAM blocks. The separate ISP and encoder blocks are not part
+  of this binding.
+
+properties:
+  compatible:
+    const: sprd,ums9230-camsys
+
+  clocks:
+    minItems: 9
+    maxItems: 9
+
+  clock-names:
+    items:
+      - const: cphy_cfg_en
+      - const: csi0
+      - const: csi1
+      - const: csi2
+      - const: mipi_csi0
+      - const: mipi_csi1
+      - const: mipi_csi2
+      - const: mipi_gate
+      - const: dcam
+
+  interrupts:
+    minItems: 3
+    maxItems: 3
+
+  interrupt-names:
+    items:
+      - const: dcam0
+      - const: dcam1
+      - const: dcam2
+
+  iommus:
+    maxItems: 1
+
+  ports:
+    $ref: /schemas/graph.yaml#/properties/ports
+
+    properties:
+      port@0: # CAMSYS_CSI_4LANE
+        $ref: /schemas/graph.yaml#/$defs/port-base
+        unevaluatedProperties: false
+        description: CSI 4-lane input port
+
+        properties:
+          endpoint:
+            $ref: video-interfaces.yaml#
+            unevaluatedProperties: false
+
+            properties:
+              data-lanes:
+                items:
+                  - const: 0
+                  - const: 1
+                  - const: 2
+                  - const: 3
+
+            required:
+              - data-lanes
+
+      port@1: # CAMSYS_CSI_M_S
+        $ref: /schemas/graph.yaml#/$defs/port-base
+        unevaluatedProperties: false
+        description: CSI M and S ports combined as 4-lane input port
+
+        properties:
+          endpoint:
+            $ref: video-interfaces.yaml#
+            unevaluatedProperties: false
+
+            properties:
+              data-lanes:
+                items:
+                  - const: 0
+                  - const: 1
+                  - const: 2
+                  - const: 3
+
+            required:
+              - data-lanes
+
+      port@2: # CAMSYS_CSI_M
+        $ref: /schemas/graph.yaml#/$defs/port-base
+        unevaluatedProperties: false
+        description: CSI 2-lane M input port
+
+        properties:
+          endpoint:
+            $ref: video-interfaces.yaml#
+            unevaluatedProperties: false
+
+            properties:
+              data-lanes:
+                items:
+                  - const: 0
+                  - const: 1
+
+            required:
+              - data-lanes
+
+      port@3: # CAMSYS_CSI_S
+        $ref: /schemas/graph.yaml#/$defs/port-base
+        unevaluatedProperties: false
+        description: CSI 2-lane S input port
+
+        properties:
+          endpoint:
+            $ref: video-interfaces.yaml#
+            unevaluatedProperties: false
+
+            properties:
+              data-lanes:
+                items:
+                  - const: 0
+                  - const: 1
+
+            required:
+              - data-lanes
+
+      port@4: # CAMSYS_CSI_RO_M_S
+        $ref: /schemas/graph.yaml#/$defs/port-base
+        unevaluatedProperties: false
+        description: CSI RO M and S ports combined as 4-lane input port
+
+        properties:
+          endpoint:
+            $ref: video-interfaces.yaml#
+            unevaluatedProperties: false
+
+            properties:
+              data-lanes:
+                items:
+                  - const: 0
+                  - const: 1
+                  - const: 2
+                  - const: 3
+
+            required:
+              - data-lanes
+
+      port@5: # CAMSYS_CSI_RO_M
+        $ref: /schemas/graph.yaml#/$defs/port-base
+        unevaluatedProperties: false
+        description: CSI 2-lane RO M input port
+
+        properties:
+          endpoint:
+            $ref: video-interfaces.yaml#
+            unevaluatedProperties: false
+
+            properties:
+              data-lanes:
+                items:
+                  - const: 0
+                  - const: 1
+
+            required:
+              - data-lanes
+
+      port@6: # CAMSYS_CSI_RO_S
+        $ref: /schemas/graph.yaml#/$defs/port-base
+        unevaluatedProperties: false
+        description: CSI 2-lane RO S input port
+
+        properties:
+          endpoint:
+            $ref: video-interfaces.yaml#
+            unevaluatedProperties: false
+
+            properties:
+              data-lanes:
+                items:
+                  - const: 0
+                  - const: 1
+
+            required:
+              - data-lanes
+
+    allOf:
+      - anyOf:
+          - required: [port@0]
+          - required: [port@1]
+          - required: [port@2]
+          - required: [port@3]
+          - required: [port@4]
+          - required: [port@5]
+          - required: [port@6]
+
+      - if:
+          required: [port@1] # CAMSYS_CSI_M_S
+        then:
+          properties:
+            port@2: false # CAMSYS_CSI_M
+            port@3: false # CAMSYS_CSI_S
+
+      - if:
+          anyOf:
+            - required: [port@2]
+            - required: [port@3]
+        then:
+          properties:
+            port@1: false
+
+      - if:
+          required: [port@4] # CAMSYS_CSI_RO_M_S
+        then:
+          properties:
+            port@5: false # CAMSYS_CSI_RO_M
+            port@6: false # CAMSYS_CSI_RO_S
+
+      - if:
+          anyOf:
+            - required: [port@5]
+            - required: [port@6]
+        then:
+          properties:
+            port@4: false
+
+  reg:
+    minItems: 7
+    maxItems: 7
+
+  reg-names:
+    items:
+      - const: csi0
+      - const: csi1
+      - const: csi2
+      - const: dcam0
+      - const: dcam1
+      - const: dcam2
+      - const: dcam_axi
+
+  resets:
+    minItems: 7
+    maxItems: 7
+
+  reset-names:
+    items:
+      - const: csi0
+      - const: csi1
+      - const: csi2
+      - const: dcam0
+      - const: dcam1
+      - const: dcam2
+      - const: dcam_axi
+
+  sprd,syscon-anlg-phy:
+    $ref: /schemas/types.yaml#/definitions/phandle
+    description: phandle of global analog syscon used for CSI PHY
+
+  sprd,syscon-mm-ahb:
+    $ref: /schemas/types.yaml#/definitions/phandle
+    description: phandle of multimedia AHB syscon
+
+required:
+  - clock-names
+  - clocks
+  - compatible
+  - interrupt-names
+  - interrupts
+  - iommus
+  - reg
+  - reg-names
+  - resets
+  - reset-names
+  - sprd,syscon-mm-ahb
+  - sprd,syscon-anlg-phy
+
+additionalProperties: false
+
+examples:
+  - |
+    #include <dt-bindings/clock/sprd,ums9230-clk.h>
+    #include <dt-bindings/interrupt-controller/arm-gic.h>
+    #include <dt-bindings/media/sprd-camsys.h>
+    #include <dt-bindings/reset/sprd,ums9230-reset.h>
+
+    camsys: camera-subsystem@3e000000 {
+      compatible = "sprd,ums9230-camsys";
+
+      reg = <0x3e200000 0x1000>,
+            <0x3e300000 0x1000>,
+            <0x3e400000 0x1000>,
+            <0x3e000000 0x1000>,
+            <0x3e001000 0x1000>,
+            <0x3e002000 0x1000>,
+            <0x3e003000 0x1000>;
+      reg-names = "csi0",
+                  "csi1",
+                  "csi2",
+                  "dcam0",
+                  "dcam1",
+                  "dcam2",
+                  "dcam_axi";
+
+      interrupts = <GIC_SPI 57 IRQ_TYPE_LEVEL_HIGH>,
+                   <GIC_SPI 58 IRQ_TYPE_LEVEL_HIGH>,
+                   <GIC_SPI 59 IRQ_TYPE_LEVEL_HIGH>;
+      interrupt-names = "dcam0", "dcam1", "dcam2";
+
+      clocks = <&aon_gate CLK_CPHY_CFG_EN>,
+               <&mm_gate CLK_MM_CSI0_EB>,
+               <&mm_gate CLK_MM_CSI1_EB>,
+               <&mm_gate CLK_MM_CSI2_EB>,
+               <&mm_clk CLK_MIPI_CSI0>,
+               <&mm_clk CLK_MIPI_CSI1>,
+               <&mm_clk CLK_MIPI_CSI2>,
+               <&mm_gate CLK_MM_MIPI_CSI0_EB>,
+               <&mm_gate CLK_MM_DCAM_EB>;
+      clock-names = "cphy_cfg_en",
+                    "csi0",
+                    "csi1",
+                    "csi2",
+                    "mipi_csi0",
+                    "mipi_csi1",
+                    "mipi_csi2",
+                    "mipi_gate",
+                    "dcam";
+
+      resets = <&mm_gate RESET_MM_AHB_MIPI_CSI0_SOFT_RST>,
+               <&mm_gate RESET_MM_AHB_MIPI_CSI1_SOFT_RST>,
+               <&mm_gate RESET_MM_AHB_MIPI_CSI2_SOFT_RST>,
+               <&mm_gate RESET_MM_AHB_DCAM0_SOFT_RST>,
+               <&mm_gate RESET_MM_AHB_DCAM1_SOFT_RST>,
+               <&mm_gate RESET_MM_AHB_DCAM2_SOFT_RST>,
+               <&mm_gate RESET_MM_AHB_DCAM_AXI_SOFT_RST>;
+      reset-names = "csi0",
+                    "csi1",
+                    "csi2",
+                    "dcam0",
+                    "dcam1",
+                    "dcam2",
+                    "dcam_axi";
+
+      sprd,syscon-mm-ahb = <&mm_ahb_regs>;
+      sprd,syscon-anlg-phy = <&anlg_phy_g10_regs>;
+      iommus = <&iommu_dcam>;
+
+      ports {
+        #address-cells = <1>;
+        #size-cells = <0>;
+
+        port@CAMSYS_CSI_4LANE {
+          reg = <CAMSYS_CSI_4LANE>;
+
+          csi_4lane_ep: endpoint {
+            data-lanes = <0 1 2 3>;
+            remote-endpoint = <&main_camera_ep>;
+          };
+        };
+      };
+    };
diff --git a/include/dt-bindings/media/sprd-camsys.h b/include/dt-bindings/media/sprd-camsys.h
new file mode 100644
index 000000000..a2378f406
--- /dev/null
+++ b/include/dt-bindings/media/sprd-camsys.h
@@ -0,0 +1,19 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+/*
+ * include/dt-bindings/media/sprd-camsys.h
+ *
+ * Copyright (C) 2025 Otto Pflüger
+ */
+
+#ifndef __DT_BINDINGS_SPRD_CAMSYS_H__
+#define __DT_BINDINGS_SPRD_CAMSYS_H__
+
+#define CAMSYS_CSI_4LANE	0
+#define CAMSYS_CSI_M_S		1
+#define CAMSYS_CSI_M		2
+#define CAMSYS_CSI_S		3
+#define CAMSYS_CSI_RO_M_S	4
+#define CAMSYS_CSI_RO_M		5
+#define CAMSYS_CSI_RO_S		6
+
+#endif /* __DT_BINDINGS_SPRD_CAMSYS_H__ */
